core/common/src/main/java/alluxio/security/authentication/ChannelAuthenticator.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.security.authentication;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.exception.status.UnauthenticatedException;
+import alluxio.grpc.SaslAuthenticationServiceGrpc;
+import alluxio.grpc.SaslMessage;
+import alluxio.util.SecurityUtils;
+import alluxio.grpc.GrpcChannelBuilder;
+
+import io.grpc.Channel;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors;
+import io.grpc.ManagedChannel;
+import io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.StreamObserver;
+
+import javax.security.auth.Subject;
+import javax.security.sasl.SaslClient;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Used to authenticate with the target host. Used internally by {@link GrpcChannelBuilder}.
+ */
+public class ChannelAuthenticator {
+
+  /** Whether to use mnarentSubject as authentication user. */
+  protected boolean mUseSubject;
+  /** Subject for authentication. */
+  protected Subject mParentSubject;
+
+  /* Used in place of a subject. */
```

Comment:
In general, can we make authentication more pluggable?

Reply:
Please see my comment on other relevant question.